internal/controller/nodetemplate_controller.go

```diff
@@ -0,0 +1,454 @@
+				"node", node.Name,
+				"nodetemplate", nodeTemplate.Name,
+			)
+			if nodeTemplate.Spec.DryRun {
+				log.Info("dry run mode, skipping update nodes.\n diff:\n"+calcDiff(nodeOriginal, &node),
+					"node", node.Name,
+					"nodetemplate", nodeTemplate.Name,
+					"phase", "updating",
+				)
+				continue
+			}
+			err = r.Client.Update(ctx, &node)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// if there are NodeTemplates that is lack of nodes with higher priority, skip ensuring of nodes.
+	nodeTemplates := nyallocatorv1.NodeTemplateList{}
+	err = r.Client.List(ctx, &nodeTemplates)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, nt := range nodeTemplates.Items {
+		if nt.Spec.Priority > nodeTemplate.Spec.Priority && !isSufficient(&nt) {
+			log.Info("skip ensuring node, because there is a NodeTemplate with higher priority that is not sufficient",
+				"nodetemplate", nodeTemplate.Name,
+				"prior", nt.Name,
+				"current", nt.Status.CurrentNodes,
+				"desired", nt.Spec.Nodes,
+			)
+			statusReason.HighPriorityNodeTemplateExists = true
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// allocate new nodes
+	allNode := corev1.NodeList{}
+	spareNodes := map[string]corev1.Node{}
+	err = r.Client.List(ctx, &allNode)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, node := range allNode.Items {
+		if _, ok := node.Labels[NodeTemplateReferenceLabelKey]; !ok && node.Labels[SpareRoleLabelKey] == "true" {
+			if selector.Matches(labels.Set(node.Labels)) {
+				spareNodes[node.Name] = node
+			}
+		}
+	}
+
+	countNodeByZone := make(map[string]int)
+	for _, n := range underManagementNodes {
+		if zone, ok := n.Labels[ZoneLabelKey]; ok {
+			countNodeByZone[zone]++
+		}
+	}
+
+	for i := len(underManagementNodes); i < nodeTemplate.Spec.Nodes; i++ {
+		if len(spareNodes) == 0 {
+			log.Info("no spare nodes found", "nodetemplate", nodeTemplate.Name)
+			statusReason.NoSpareNodesFound = true
+			return ctrl.Result{}, nil
+		}
+
+		candidate := selectNodes(spareNodes, countNodeByZone)
+		delete(spareNodes, candidate.Name)
+		if zone, ok := candidate.Labels[ZoneLabelKey]; ok {
+			countNodeByZone[zone]++
+		}
+		candidateOriginal := candidate.DeepCopy()
+
+		r.reflectNodeConfiguration(&candidate, nodeTemplate)
+		if nodeTemplate.Spec.DryRun {
+			log.Info("dry run mode, skipping update nodes.\n diff:\n"+calcDiff(candidateOriginal, &candidate),
+				"node", candidate.Name,
+				"nodetemplate", nodeTemplate.Name,
+				"phase", "ensuring",
+			)
+			continue
+		}
+		err := r.Client.Update(ctx, &candidate)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("allocated node",
+			"node", candidate.Name,
+			"nodetemplate", nodeTemplate.Name,
+		)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NodeTemplateReconciler) reflectNodeConfiguration(node *corev1.Node, nodeTemplate *nyallocatorv1.NodeTemplate) {
+	if nodeTemplate.Spec.Template.Metadata.Labels != nil && node.Labels == nil {
```

Comment:
This nil check(`nodeTemplate.Spec.Template.Metadata.Labels != nil`) is unnecessary. 
NodeTemplateReconciler always adds labels regardless of the NodeTemplate's spec.

```suggestion
	if node.Labels == nil {
```